grammarinator/generate.py

```diff
@@ -126,12 +121,12 @@ def create_new_test(self, index, lock):
             logger.warning('Test generation failed.', exc_info=e)
             return self.create_new_test(index, lock)
 
-        test_fn = self.out_format % index if self.out_format else None
+        test_fn = self.out_format % index if '%d' in self.out_format else self.out_format
         tree.root = Generator.transform(tree.root, self.transformers)
 
         tree_fn = None
         if self.population and self.keep_trees:
-            tree_fn = join(self.population.directory, basename(test_fn) + Tree.extension)
+            tree_fn = join(self.population.directory, (basename(test_fn) if test_fn else str(index)) + Tree.extension)
```

Comment:
I'm sorry, my previous suggestion wasn't perfect either. We have to save  trees separately even if all the tests are saved under the same name. So, we have to do something similar here:

```py
if test_fn:
    tree_basename = basename(test_fn)
    if '%d' not in tree_basename:
        base, ext = splitext(tree_basename)
        tree_basename = '{base}%d{ext}'.format(base=base, ext=ext)
else:
    tree_basename = '%d'
tree_fn = join(self.population.directory, tree_basename % index + Tree.extension)
```